core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -657,6 +657,9 @@ class GroupMetadataManager(brokerId: Int,
             }
             currOffset = batch.nextOffset
           }
+          if (currOffset == log.logStartOffset) {
```

Comment:
Another way we could handle this is to log a warning and break from the loop. The problem is that the cause of the inconsistent start and end offsets is somewhere else. So failing here may not actually help us determine the root cause. And what should a user do if they hit this? The problem is that the broker doesn't actually fail, it just fails to load the partition, so an operator may not even notice the failure until someone complains about consumer behavior.

Reply:
Indeed, we could log a warning and break from the loop. I went with an `IllegalStateException` to stay consistent with the other errors thrown in the method but I don't have a strong preference for one or the other.

You're right. The cause is somewhere else and the broker won't fail. I think the best we can do is to make the loading robust to such situation and log an error to allow the operator to troubleshoot the issue. Deleting the empty segment(s) solves the issue.

Now that I am thinking about it, another way to solve this would be to ensure that empty segments are cleanup in the LogCleaner. I believe that this is the only reasonable way which could lead to an empty segment and it sounds more appropriate. I will investigate.

Thanks for the comment.

Reply:
I think the empty segment is not so much the issue here, but the fact that the start and end offsets have gotten out of sync with each other. If the log is empty, then they should match each other. I'm not too sure how this can happen.. Manual deletion is probably one way, but it wouldn't surprise to find bugs somewhere in the log layer that could cause it also. The main question is whether we should allow bugs in the log to propagate to failures here and honestly I'm not sure the answer. Ignoring state inconsistencies can also be dangerous.

Reply:
I am not sure to understand your point regarding the offsets that have gotten out of sync. I have observed the following on an impacted broker:
```
     0B Aug 17 14:45 00000000000000000000.index
     0B Aug 17 14:45 00000000000000000000.log
     0B Aug 17 14:45 00000000000000000000.timeindex
    10M Oct  9 23:17 00000000000010342848.index
     0B Aug 18 11:13 00000000000010342848.log
    10B Aug 17 14:45 00000000000010342848.snapshot
    10M Oct  9 23:17 00000000000010342848.timeindex
```
When the partition is loaded, start offset is 0 and end offset is 10342848 which sounds reasonable to me. Should we observe something different in this case?

The operator of that cluster has assured me that no manual interventions have been made on the partition.

I do agree that ignoring state inconsistencies is dangerous. The best would be the find the underlying conditions which have led to this.

Reply:
@dajac After thinking about this some more, I think we need to make the loading logic resilient to empty segments. There are several legitimate reasons this can happen. For example, it is definitely possible that the active segment can be empty if the broker crashes after rolling. The cleaner may also leave empty segments in its current handling due to limitations on the range of offsets representable in a single index file. I think the main problem with this code is that it assumes that we can reach the log end offset by starting from the beginning and incrementing the offset of the next fetched record, but I don't see how we can guarantee this in general. I think my suggestion would be to break from the loop if the call to `Log.read` returns an empty record set.

Reply:
@hachikuji Thanks for coming back to this one. I agree that it makes sense to make the code more robust here. I have reworked the PR based on your suggestion. It is the first time that I use `break` in Scala so I am not sure that it is a best practice. Please let me know what do you think.